Docker

jseliga opened the pull request:
Vytvořil jsem základní konfiguraci pro docker.

Co zatím stále dělá problém je mongo ‒ nedaří se mi k ní připojit z aplikace. Můžete na to prosím mrknout @lotcz @alutom?

Commit:
Added Dockerfile and docker-compose.yml

Commit:
Added docs for docker

Commit:
Timetracking

alutom committed:
Run in docker without tomcat

alutom commented:
Ahoj, jsou v zásadě dva způsoby jak tu aplikaci pustit v dockeru:
1. Pustit přímo jar přes `java -jar ...`
2. Deployment na aplikační server (např. tomcat)

Tvůj Dockerfile obsahuje oba způsoby, ale ani jeden nemůže fungovat, protože pro způsob č. 1 je potřeba buildit jar místo waru a pro způsob č. 2 musí aplikace rozšiřovat třídu `SpringBootServletInitializer`. Tady jsou obě varianty opravené:
1. jar build + java: bcb1d5d7ee54d8d734667606b58017449362beae
2. war build + tomcat: 2a64f8e257ce2824128315f9f9f6a1e889bf624d

Asi je na nás, který způsob zvolíme. Možná je oba možné nějak jednoduše sloučit, aby mezi nima šlo přepínat. Zas tak moc tomu nerozumím.

BTW, také jsem zjistil, že po provedení změn musím pustit `mvnw package && docker-compile up --build`. Bez parametru `--build` se použije docker image z předchozího spuštění.

lotcz commented:
Ahoj,

mohli byste prosím přidat návod pro lamy, jako jsem já? Docker mi vytvoří několik image - java, mongo a web_src. Nevím, jestli se mají spustit všechny, nebo jen ten web. Podařilo se mi po chvíli snažení rozběhnout aplikaci, ale nefunguje připojení na Mongo.

Dále jsem si všiml, že se tam operuje s hostname ubuntu-server.loc - to je jen název virtuálu, na kterém mám Mongo u sebe na počítači. To bych tam nedával - v Dockeru by vše mělo běžet v jednom kontejneru (na jednom stroji), takže localhost. Potom Mongo umožňuje připojit se bez uživatelského jména a hesla, pokud aplikace běží lokálně - to taky chceme.

V souboru application.properties by tedy podle mělo stačit jen toto:
```
spring.data.mongodb.database=liquor
```
vše ostatní je v defaultu a uživatel není potřeba

alutom commented:
Ahoj,

od toho je tam právě soubor `docker-compose.yml`, který popisuje jaké kontejnery vytvořit a jak je sesíťovat dohromady. Mně to třeba funguje v branchi `docker-war`. Tam prostě pustím tento příkaz a vše se rozběhne:
```
mvnw package && docker-compose up --build
```
Funguje to s těmi `application.properties`, které jsou v repu, ale asi to tedy ještě změníme. V předchozím příspěvku jsem zmiňoval `docker-compile`, to byl omyl.

jseliga commented:
Díky @alutom.
@lotcz, je pro nás lepší build do waru a nebo do jaru? Preferujeme něco?

lotcz commented:
> Díky @alutom.
> @lotcz, je pro nás lepší build do waru a nebo do jaru? Preferujeme něco?

mě přijde jednodušší jar, když na nic jiného vlastně Tomcat nepořebujeme, ale nijak zvlášť silný názor na to nemám.

lotcz commented:
> Ahoj,
> 
> od toho je tam právě soubor `docker-compose.yml`, který popisuje jaké kontejnery vytvořit a jak je sesíťovat dohromady. Mně to třeba funguje v branchi `docker-war`. Tam prostě pustím tento příkaz a vše se rozběhne:
> 
> ```
> mvnw package && docker-compose up --build
> ```
> 
> Funguje to s těmi `application.properties`, které jsou v repu, ale asi to tedy ještě změníme. V předchozím příspěvku jsem zmiňoval `docker-compile`, to byl omyl.

no mě to nefunguje nejspíš proto, že ubuntu-server.loc u mě už existuje. Taky jsem zkoušel jenom tu variantu s jar.

A je to tedy tak, že celá aplikace je v tom image web_src? Nebo je mongo zvlášť a musí se tedy spustit oba dva kontejnery?

díky

Karel

alutom commented:
@lotcz Ano, nejspíš ti to nefunguje kvůli té kolizi. Dáme tam stejně něco jiného, protože nedává smysl ten hostname používat, jak jsi už ostatně naznačil dřív.

Aktuálně je to tak, že tam jsou dva kontejnery - jeden s naší aplikací a jeden s mongodb. Musí být puštěné oba. Chápu správně, že tedy chceme jeden, kde bude všechno? Pokud je to definitivní rozhodnutí, nabízím se, že to udělám, ale pokud v tom chce pokračovat @jseliga, rád mu to přenechám.

Osobně tedy navrhuji war, protože tím nic neztratíme, ale možnost nasadit to na tomcat se může hodit.

lotcz commented:
Myslím, že všechny varianty jsou životaschopné a mají své výhody. Navíc se to poměrně snadno dá později změnit, takže bych se tím nějak moc nezdržoval.

Osobně bych volil variantu jednoho kontejneru s buildem do jar - pro naše účely školního projektu je to podle mě cesta nejmenšího odporu.

Asi to bude muset rozhodnout @zdebou - jar nebo war, jeden kontejner nebo dva?

jseliga commented:
@alutom Ok, klidně se do toho dej.

Dockerfile se může případně udělat s přepínačem, jestli má brát jar/war. Ale je pravda, že jar bude stačit.

zdebou commented:
Z mého pohledu je důležité jednoduché nasazení bez řešení dns, apod. (kubernetes, apod.) => z toho mi nejlépe vychází jeden kontejner. Pro vývoj jsou ale lepší asi dva, jeden s funkční DB a k tomu debug.

alutom committed:
Create single container with app and db

alutom committed:
Merge branch 'master' into docker

alutom committed:
Timetracking

alutom commented:
Upravil jsem tedy tento pull request. Build se provádí do jaru a mongo je ve stejném kontejneru jako aplikace.

alutom committed:
Update instructions for Docker

Commit:
- added executable bit to scripts
- removed mongo user and host/port (using defaults)
- application.properties now work with docker
- added example config
